### PR TITLE
More changes to fix uneven texture maximums

### PR DESCRIPTION
**Pull request by Varen**
image
[Link bug 2](https://youtu.be/cCoSve4mS2w)

Drawing on textures with some combinations of strengths causes the total to go above 255, as well as some cases of swapping and other stuff that rounds texture values. This is apparent only in game when it happens, as noggit doesn't seem to care to render the problem. It does however show up if some operations are done on the textures.

I added checks to all the paths I found this can happen, could be a few more I haven't found though. 
Won't solve it on load for pre-fix adts but that'll be fixed by doing something on it, like swapping affected chunks textures around.

### CODE REVIEW

**Comment by Varen** [comment]
Did a new PR as the old one didn't fix all cases.

**Comment by bloerwald** [comment]
> Did a new PR as the old one didn't fix all cases.

FYI: No need to create a new PR, you can just push to the old branch and it will automatically update. :)

File: src/noggit/texture_set.cpp [hunk 1]
```diff
@@ -425,11 +426,40 @@ bool TextureSet::paintTexture(float xbase, float zbase, float x, float z, Brush*
...
+
+            if (amaps[layer][offset] > 1.0f)
+                amaps[layer][offset] = 1.0f;
+            else if (amaps[layer][offset] < 0.0f || isnan(amaps[layer][offset]))
+                    amaps[layer][offset] = 0.0f;
+
+            if(layer != 0)
+                totOthers += amaps[layer][offset];
+          }
+          if (totOthers > 1)
```
Comment (bloerwald) [inline src/noggit/texture_set.cpp:440]: `> 1` or `> 0`? I would intuitively assume `> 0`?

File: src/noggit/texture_set.cpp [hunk 2]
```diff
@@ -425,11 +426,40 @@ bool TextureSet::paintTexture(float xbase, float zbase, float x, float z, Brush*
             else
             {
               amaps[layer][offset] -= alpha_change * (amaps[layer][offset] / sum_other_alphas);
+
+            }
+
+            if (amaps[layer][offset] > 1.0f)
+                amaps[layer][offset] = 1.0f;
+            else if (amaps[layer][offset] < 0.0f || isnan(amaps[layer][offset]))
```
Comment (bloerwald) [inline src/noggit/texture_set.cpp:434]: What is the operation making it nan, ` / sum_other_alphas`? If that's it, I'd suggest changing `sum_other_alphas = 1.f - current_alpha` above to `sum_other_alphas = std::max (1.f - current_alpha, std::numeric_limits<float>::epsilon())` (include `<limits>` if that's not yet the case), which would make the division always sane and never produce a NaN.

Then, here we can just `amaps[layer][offset] = std::min (1.f, std::max (0.f, amaps[layer][offset]));` to express what the condition is.

File: src/noggit/texture_set.cpp [hunk 3]
```diff
@@ -425,11 +426,40 @@ bool TextureSet::paintTexture(float xbase, float zbase, float x, float z, Brush*
...
+                    amaps[layer][offset] = 0.0f;
+
+            if(layer != 0)
+                totOthers += amaps[layer][offset];
+          }
+          if (totOthers > 1)
+          {
+            float mul = (1 / totOthers);
+
+            amaps[1][offset] = amaps[1][offset] * mul;
```
Comment (bloerwald) [inline src/noggit/texture_set.cpp:444]: This one isn't guarded, but probably is implicitly guarded by the condition only happening if `nTextures > 0`. The loop

```
auto corrected_totOthers = 0.f;
for (auto layer (1); layer < nTextures; ++layer) { 
  amaps[layer][offset] = amaps[layer][offset] / totOthers;
  corrected_totOthers += amaps[layer][offset];
}
amaps[0][offset] = std::max(0.0f, std::min(1.0f - corrected_totOthers, 1.0f));
```
should be equivalent?

File: src/noggit/texture_set.cpp [hunk 4]
```diff
@@ -931,6 +961,9 @@ uint8_t TextureSet::sum_alpha(size_t offset) const
       sum += amap->getAlpha(offset);
     }
   }
+  // This can happen with some previous version saved adts.
+  if (sum > 255)
+      sum = 255;
```
Comment (bloerwald) [inline src/noggit/texture_set.cpp:966]:
```suggestion
  sum = std::min (sum, std::numeric_limits<uint8_t>::max(), sum);
```

File: src/noggit/texture_set.cpp [hunk 5]
```diff
@@ -946,24 +979,35 @@ namespace
...
-    }
+        for (int alpha_layer = 0; alpha_layer < nTextures - 1; ++alpha_layer)
+        {
+            values[alpha_layer][i] = float_alpha_to_uint8(new_amaps[alpha_layer + 1][i] * 255.f);
+        }
+        float totVal = (float)(values[0][i]) + values[1][i] + values[2][i];
+        if (totVal > 255)
+        {
+            float mul = (255 / totVal);
+            float vals[3] = { std::floorf(values[0][i] *mul), std::floorf(values[1][i]*mul), std::floorf(values[2][i] *mul)};
```
Comment (bloerwald) [inline src/noggit/texture_set.cpp:1000]: I was slightly confused by this array, so I'd personally just inline it: the values and following lines are independent, so `v = fatu (floor (v  * mul))` seems easier to read.

**Comment by bloerwald** [comment]
This all looks fine, **but** I'm slightly scared changing this. I know that in WoD they changed the blending formula, but I don't know whether they did it between BC and WotLK. If they did, this might be correct for WotLK (big alpha) but broken for maps that don't use that yet (either little alpha or even "unfixed" alpha). Can you do me a favor and also test this with some BC map, i.e. Expansion01, which iirc wasn't updated to big alpha?

**Comment by Varen** [comment]
Looks like this is all fixed with the latest release in better ways, so gunna close this!

**Pull request closed without merging**